edb/server/protocol/request_scheduler.py

```diff
@@ -0,0 +1,570 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+from dataclasses import dataclass, field
+from typing import (
+    Final,
+    Generic,
+    Iterable,
+    Literal,
+    Optional,
+    Sequence,
+    TypeVar,
+)
+
+import abc
+import asyncio
+import copy
+import random
+
+_T = TypeVar('_T')
+
+
+@dataclass
+class Timer:
+    """Represents a time after which an action should be taken.
+
+    Examples:
+    (None, True) = execute immediately
+    (None, False) = execute any time
+    (10, True) = execute immediately after 10s
+    (10, False) = execute any time after 10s
+    """
+
+    time: Optional[float] = None
+
+    # Whether the action should be taken as soon as possible after the time.
+    urgent: bool = True
+
+    @staticmethod
+    def create_delay(delay: Optional[float], urgent: bool):
+        now = asyncio.get_running_loop().time()
+        if delay is None:
+            time = None
+        else:
+            now = asyncio.get_running_loop().time()
+            time = now + delay
+
+        return Timer(time, urgent)
+
+    def is_ready(self) -> bool:
+        now = asyncio.get_running_loop().time()
+        return self.time is None or self.time <= now
+
+    def is_ready_and_urgent(self) -> bool:
+        return self.is_ready() and self.urgent
+
+    def remaining_time(self, max_delay: float) -> float:
+        """How long before this timer is ready in seconds."""
+        if self.urgent:
+            if self.time is None:
+                return 0
+
+            else:
+                # 1ms extra, just in case
+                now = asyncio.get_running_loop().time()
+                delay = self.time - now + 0.001
+                return min(max(0, delay), max_delay)
+
+        else:
+            # If not urgent, wait as long as possible
+            return max_delay
+
+    @staticmethod
+    def combine(timers: Iterable[Timer]) -> Timer:
+        """Combine the timers to determine the when to take the next action.
+
+        If the timers are (1, False), (2, False), (3, True), it may be wasteful
+        to act at times [1, 2, 3].
+
+        Instead, we would prefer to act only once, at time 3, since only the
+        third action was urgent.
+        """
+        if any(
+            timer.time is None and timer.urgent
+            for timer in timers
+        ):
+            # An action must be taken right away.
+            return Timer(None, True)
+
+        urgent_times = [
+            timer.time
+            for timer in timers
+            if timer.time is not None and timer.urgent
+        ]
+        if len(urgent_times) > 0:
+            # An action must be taken after some delay
+            return Timer(min(urgent_times), True)
+
+        # Providers can be processed whenever
+        return Timer(None, False)
```

Comment:
Since non-urgent timers can be handled whenever, a user of the class is free to do whatever they like.

I suppose it could be meaningful to return the highest timer, but otherwise, it seems unnecessary.

Reply:
They can't be handled before the timer though, right?

Reply:
Thinking about it a bit more, it's probably best to pass a time anyways and let the caller deal with it.